backend/postgres/postgres.go

```diff
@@ -0,0 +1,955 @@
+		}
+
+		existingEvents = append(existingEvents, e)
+	}
+
+	state := backend.NewOrchestrationRuntimeState(wi.InstanceID, existingEvents)
+	return state, nil
+}
+
+// GetOrchestrationWorkItem implements backend.Backend
+func (be *postgresBackend) GetOrchestrationWorkItem(ctx context.Context) (*backend.OrchestrationWorkItem, error) {
+	if err := be.ensureDB(ctx); err != nil {
+		return nil, err
+	}
+
+	tx, err := be.db.BeginTx(ctx, nil)
+	if err != nil {
+		return nil, err
+	}
+	defer tx.Rollback()
+
+	now := time.Now().UTC()
+	newLockExpiration := now.Add(be.orchestrationLockTimeout)
+
+	// Place a lock on an orchestration instance that has new events that are ready to be executed.
+	row := tx.QueryRowContext(
+		ctx,
+		`UPDATE durabletask.Instances SET LockedBy = $1, LockExpiration = $2
+		WHERE ctid = (
+			SELECT ctid FROM durabletask.Instances I
+			WHERE (I.LockExpiration IS NULL OR I.LockExpiration < $3) AND EXISTS (
+				SELECT 1 FROM durabletask.NewEvents E
+				WHERE E.InstanceID = I.InstanceID AND (E.VisibleTime IS NULL OR E.VisibleTime < $4)
+			)
+			LIMIT 1
+		) RETURNING InstanceID`,
+		be.workerName,     // LockedBy for Instances table
+		newLockExpiration, // Updated LockExpiration for Instances table
+		now,               // LockExpiration for Instances table
+		now,               // VisibleTime for NewEvents table
+	)
+
+	if err := row.Err(); err != nil {
+		if ctx.Err() != nil {
+			return nil, ctx.Err()
+		}
+		return nil, fmt.Errorf("failed to query for orchestration work-items: %w", err)
+	}
+
+	var instanceID string
+	if err := row.Scan(&instanceID); err != nil {
+		if err == sql.ErrNoRows {
+			// No new events to process
+			return nil, backend.ErrNoWorkItems
+		}
+
+		return nil, fmt.Errorf("failed to scan the orchestration work-item: %w", err)
+	}
+
+	// TODO: Get all the unprocessed events associated with the locked instance
+	events, err := tx.QueryContext(
+		ctx,
+		`WITH selected AS (
+			UPDATE durabletask.NewEvents SET DequeueCount = DequeueCount + 1, LockedBy = $1 WHERE ctid IN (
+				SELECT ctid FROM durabletask.NewEvents
+				WHERE InstanceID = $2 AND (VisibleTime IS NULL OR VisibleTime <= $3)
+				LIMIT 1000
+			) RETURNING EventPayload, DequeueCount, SequenceNumber
+		)
+		SELECT EventPayload, DequeueCount FROM selected ORDER BY SequenceNumber ASC`,
+		be.workerName,
+		instanceID,
+		now,
+	)
+	if err != nil {
+		if ctx.Err() != nil {
+			return nil, ctx.Err()
+		}
+		return nil, fmt.Errorf("failed to query for orchestration work-items: %w", err)
+	}
+
+	maxDequeueCount := int32(0)
+
+	newEvents := make([]*protos.HistoryEvent, 0, 10)
+	for events.Next() {
+		var eventPayload []byte
+		var dequeueCount int32
+		if err := events.Scan(&eventPayload, &dequeueCount); err != nil {
+			return nil, fmt.Errorf("failed to read history event: %w", err)
+		}
+
+		if dequeueCount > maxDequeueCount {
+			maxDequeueCount = dequeueCount
+		}
+
+		e := new(protos.HistoryEvent)
+		err := protojson.Unmarshal(eventPayload, e)
```

Comment:
Why not storing protobuf inside the DB instead? It's more efficient than JSON anyways, as long as we don't care that users may want to modify the data themselves by hand?

Reply:
Debuggability. It's incredibly painful to try and inspect the database when things go wrong if we store everything as protobuf (which is what we do currently in Dapr and in the SQLite database). I thought the performance vs. debuggability tradeoff would be worthwhile.

Reply:
That can make sense. I would still recommend changing the column types to TEXT however, unless you need to perform queries that look into the JSON data.

Reply:
Or... https://github.com/mpartel/postgres-protobuf :)